In [ ]:
#hide_output
%load_ext autoreload
%autoreload 2

# Large Git repositories

Imports for handling graphs in Python

In [ ]:
import networkx as nx

Imports for cloning Git repositories and extracting the commit graph

In [ ]:
from git_commit_graph_ext.commit_graph import commit_graph, _repo_graph_name

Gathering cloned repositories and their graphs

In [ ]:
known_repos = []

This should perhaps be a list of objects of specific defined type, not list of dicts

In [ ]:
def get_known_repos():
    return known_repos

In [ ]:
get_known_repos()

[]

## From `gen-test` repository

[gen-test](https://github.com/derrickstolee/gen-test) -- Test scripts for testing new versions of generation numbers.

From the README:

> This report investigates four replacements for generation numbers, and compares the number of walked commits to the existing algorithms (both using generation numbers and not using them at all). We can use this data to make decisions for the future of the feature.

https://github.com/derrickstolee/gen-test/blob/master/clone-repos.sh

```sh
git clone git@github.com:derrickstolee/git.git --branch reach-perf

git clone git@github.com:curl/curl
git clone git@github.com:electron/electron
git clone git@github.com:FFmpeg/FFmpeg
git clone git@github.com:eclipse/jgit
git clone git@github.com:JuliaLang/julia
git clone git@github.com:jetbrains/kotlin
git clone git@github.com:torvalds/linux
git clone git@github.com:odoo/odoo
git clone git@github.com:openssl/openssl
git clone git@github.com:apple/swift
git clone git@github.com:tensorflow/tensorflow
git clone git@github.com:microsoft/TypeScript

git clone https://gerrit.googlesource.com/gerrit
git clone https://android.googlesource.com/platform/frameworks/base/ android-base
git clone https://chromium.googlesource.com/chromium/src chromium
```

> Note: the repositoriess from googlesource.com have trouble with cloning; it might be better to clone them from command line, not from this Jupyter notebook.

### cURL repository

Curl is a command-line tool for transferring data specified with URL syntax.

In [ ]:
%%time
curl_graph = commit_graph('https://github.com/curl/curl.git', 'curl')

Wall time: 9.6 s


It takes around 9 s to reclone the `curl` repository, 4 s to rescan it, and 300 ms to refresh it.

In [ ]:
known_repos = []
known_repos.append(
    {
        'name': 'curl',
        'url': 'https://github.com/curl/curl',
        'graph': curl_graph,
    }
)
known_repos[0]

{'name': 'curl',
 'url': 'https://github.com/curl/curl',
 'graph': <networkx.classes.digraph.DiGraph at 0x14d549b1c70>}

In [ ]:
print('commit graph of "{}" project has'.format(known_repos[0]['name']))
print('- {} nodes / vertices / commits'.format(known_repos[0]['graph'].number_of_nodes()))
print('- {} edges'.format(known_repos[0]['graph'].number_of_edges()))
print('- {:.5f} edge density, number of edges divided by number of nodes (vertices)'.
      format(known_repos[0]['graph'].number_of_edges() / known_repos[0]['graph'].number_of_nodes()))
print('- {} sinks / leafs / roots in git parlance, nodes with no outgoing edges'.
     format(len([n for n in known_repos[0]['graph'] if known_repos[0]['graph'].out_degree(n) == 0])))
print('- {} sources / roots / heads in git parlance, nodes with no incoming edge'.
     format(len([n for n in known_repos[0]['graph'] if known_repos[0]['graph'].in_degree(n) == 0])))

commit graph of "curl" project has
- 26609 nodes / vertices / commits
- 26631 edges
- 1.00083 edge density, number of edges divided by number of nodes (vertices)
- 1 sinks / leafs / roots in git parlance, nodes with no outgoing edges
- 14 sources / roots / heads in git parlance, nodes with no incoming edge


### Electron repository

Electron -- Build cross-platform desktop apps with JavaScript, HTML, and CSS

In [ ]:
%%time
electron_graph = commit_graph('https://github.com/electron/electron.git', 'electron')

Wall time: 892 ms


It takes around 15 seconds to clone the Electron repository, and construct it's commit graph as NetworkX.DiGraph

In [ ]:
known_repos.append({
    'name': 'electron',
    'url': 'https://github.com/electron/electron.git',
    'graph': electron_graph,
})
known_repos

[{'name': 'curl',
  'url': 'https://github.com/curl/curl',
  'graph': <networkx.classes.digraph.DiGraph at 0x14d549b1c70>},
 {'name': 'electron',
  'url': 'https://github.com/electron/electron.git',
  'graph': <networkx.classes.digraph.DiGraph at 0x14d549b17f0>}]

In [ ]:
print('commit graph of "{}" project has'.format(known_repos[1]['name']))
print('- {} nodes / vertices / commits'.format(known_repos[1]['graph'].number_of_nodes()))
print('- {} edges'.format(known_repos[1]['graph'].number_of_edges()))
print('- {:.5f} edge density, number of edges divided by number of nodes (vertices)'.
      format(known_repos[1]['graph'].number_of_edges() / known_repos[1]['graph'].number_of_nodes()))
print('- {} sinks / leafs / roots in git parlance, nodes with no outgoing edges'.
     format(len([n for n in known_repos[1]['graph'] if known_repos[1]['graph'].out_degree(n) == 0])))
print('- {} sources / roots / heads in git parlance, nodes with no incoming edge'.
     format(len([n for n in known_repos[1]['graph'] if known_repos[1]['graph'].in_degree(n) == 0])))

commit graph of "electron" project has
- 29956 nodes / vertices / commits
- 34167 edges
- 1.14057 edge density, number of edges divided by number of nodes (vertices)
- 4 sinks / leafs / roots in git parlance, nodes with no outgoing edges
- 142 sources / roots / heads in git parlance, nodes with no incoming edge


### FFmpeg repository

<https://github.com/FFmpeg/FFmpeg> is a mirror of <https://git.ffmpeg.org/ffmpeg.git>

FFmpeg is a collection of libraries and tools to process multimedia content such as audio, video, subtitles and related metadata.

In [ ]:
%%time
ffmpeg_graph = commit_graph('https://github.com/FFmpeg/FFmpeg.git', 'ffmpeg')

Wall time: 2.05 s


It takes around 14 seconds to clone the FFmpeg repository, and construct it's commit graph as NetworkX.DiGraph,  
and around 1.4 seconds to refresh it.

In [ ]:
known_repos.append({
    'name': 'ffmpeg',
    'url': 'https://github.com/FFmpeg/FFmpeg.git',
    'graph': ffmpeg_graph
})
known_repos

[{'name': 'curl',
  'url': 'https://github.com/curl/curl',
  'graph': <networkx.classes.digraph.DiGraph at 0x14d549b1c70>},
 {'name': 'electron',
  'url': 'https://github.com/electron/electron.git',
  'graph': <networkx.classes.digraph.DiGraph at 0x14d549b17f0>},
 {'name': 'ffmpeg',
  'url': 'https://github.com/FFmpeg/FFmpeg.git',
  'graph': <networkx.classes.digraph.DiGraph at 0x14d549b1e50>}]

In [ ]:
graph_idx = 2
print('commit graph [{}] of "{}" project has'.format(graph_idx, known_repos[graph_idx]['name']))
print('- {} nodes / vertices / commits'.format(known_repos[graph_idx]['graph'].number_of_nodes()))
print('- {} edges'.format(known_repos[graph_idx]['graph'].number_of_edges()))
print('- {:.5f} edge density, number of edges divided by number of nodes (vertices)'.
      format(known_repos[graph_idx]['graph'].number_of_edges() / known_repos[graph_idx]['graph'].number_of_nodes()))
print('- {} sinks / leafs / roots in git parlance, nodes with no outgoing edges'.
     format(len([n for n in known_repos[graph_idx]['graph'] if known_repos[graph_idx]['graph'].out_degree(n) == 0])))
print('- {} sources / roots / heads in git parlance, nodes with no incoming edge'.
     format(len([n for n in known_repos[graph_idx]['graph'] if known_repos[graph_idx]['graph'].in_degree(n) == 0])))

commit graph [2] of "ffmpeg" project has
- 120583 nodes / vertices / commits
- 130606 edges
- 1.08312 edge density, number of edges divided by number of nodes (vertices)
- 1 sinks / leafs / roots in git parlance, nodes with no outgoing edges
- 30 sources / roots / heads in git parlance, nodes with no incoming edge


### JGit repository

JGit - An implementation of the Git version control system in pure Java

In [ ]:
%%time
jgit_graph = commit_graph('https://github.com/eclipse/jgit.git', 'jgit')

Wall time: 3.41 s


It takes around 34 seconds to clone the JGit repository, and construct it's commit graph as NetworkX.DiGraph,  
and 70 ms to refresh it

In [ ]:
#known_repos[3] = {
known_repos.append({
    'name': 'jgit',
    'url': 'https://github.com/eclipse/jgit.git',
    'graph': jgit_graph
})
known_repos

[{'name': 'curl',
  'url': 'https://github.com/curl/curl',
  'graph': <networkx.classes.digraph.DiGraph at 0x14d549b1c70>},
 {'name': 'electron',
  'url': 'https://github.com/electron/electron.git',
  'graph': <networkx.classes.digraph.DiGraph at 0x14d549b17f0>},
 {'name': 'ffmpeg',
  'url': 'https://github.com/FFmpeg/FFmpeg.git',
  'graph': <networkx.classes.digraph.DiGraph at 0x14d549b1e50>},
 {'name': 'jgit',
  'url': 'https://github.com/eclipse/jgit.git',
  'graph': <networkx.classes.digraph.DiGraph at 0x14d61fc1280>}]

On 01.01.2021 the JGit project had 8,162 commits, according to GitHub web UI

In [ ]:
graph_idx = 3
print('commit graph [{}] of "{}" project has'.format(graph_idx, known_repos[graph_idx]['name']))
print('- {} nodes / vertices / commits'.format(known_repos[graph_idx]['graph'].number_of_nodes()))
print('- {} edges'.format(known_repos[graph_idx]['graph'].number_of_edges()))
print('- {:.5f} edge density, number of edges divided by number of nodes (vertices)'.
      format(known_repos[graph_idx]['graph'].number_of_edges() / known_repos[graph_idx]['graph'].number_of_nodes()))
print('- {} sinks / leafs / roots in git parlance, nodes with no outgoing edges'.
     format(len([n for n in known_repos[graph_idx]['graph'] if known_repos[graph_idx]['graph'].out_degree(n) == 0])))
print('- {} sources / roots / heads in git parlance, nodes with no incoming edge'.
     format(len([n for n in known_repos[graph_idx]['graph'] if known_repos[graph_idx]['graph'].in_degree(n) == 0])))

commit graph [3] of "jgit" project has
- 8173 nodes / vertices / commits
- 9853 edges
- 1.20555 edge density, number of edges divided by number of nodes (vertices)
- 1 sinks / leafs / roots in git parlance, nodes with no outgoing edges
- 2 sources / roots / heads in git parlance, nodes with no incoming edge


### The Julia language repository

The Julia Programming Language  
<https://julialang.org/>

Julia is a high-level, high-performance dynamic language for technical computing.

In [ ]:
%%time
julia_graph = commit_graph('https://github.com/JuliaLang/julia.git', 'julia')

Wall time: 1.04 s


It takes around 32.7 seconds to run `commit_graph()` from scratch

The Julia language repository had 48,658 commits on 01.01.2021

In [ ]:
known_repos.append({
    'name': 'julia',
    'url': 'https://github.com/JuliaLang/julia.git',
    'homepage': 'https://julialang.org',
    'graph': julia_graph
})
known_repos

[{'name': 'curl',
  'url': 'https://github.com/curl/curl',
  'graph': <networkx.classes.digraph.DiGraph at 0x14d549b1c70>},
 {'name': 'electron',
  'url': 'https://github.com/electron/electron.git',
  'graph': <networkx.classes.digraph.DiGraph at 0x14d549b17f0>},
 {'name': 'ffmpeg',
  'url': 'https://github.com/FFmpeg/FFmpeg.git',
  'graph': <networkx.classes.digraph.DiGraph at 0x14d549b1e50>},
 {'name': 'jgit',
  'url': 'https://github.com/eclipse/jgit.git',
  'graph': <networkx.classes.digraph.DiGraph at 0x14d61fc1280>},
 {'name': 'julia',
  'url': 'https://github.com/JuliaLang/julia.git',
  'homepage': 'https://julialang.org',
  'graph': <networkx.classes.digraph.DiGraph at 0x14d61fc1d30>}]

In [ ]:
graph_idx = 4
print('commit graph [{}] of "{}" project has'.format(graph_idx, known_repos[graph_idx]['name']))
print('- {} nodes / vertices / commits'.format(known_repos[graph_idx]['graph'].number_of_nodes()))
print('- {} edges'.format(known_repos[graph_idx]['graph'].number_of_edges()))
print('- {:.5f} edge density, number of edges divided by number of nodes (vertices)'.
      format(known_repos[graph_idx]['graph'].number_of_edges() / known_repos[graph_idx]['graph'].number_of_nodes()))
print('- {} sinks / leafs / roots in git parlance, nodes with no outgoing edges'.
     format(len([n for n in known_repos[graph_idx]['graph'] if known_repos[graph_idx]['graph'].out_degree(n) == 0])))
print('- {} sources / roots / heads in git parlance, nodes with no incoming edge'.
     format(len([n for n in known_repos[graph_idx]['graph'] if known_repos[graph_idx]['graph'].in_degree(n) == 0])))

commit graph [4] of "julia" project has
- 56048 nodes / vertices / commits
- 67547 edges
- 1.20516 edge density, number of edges divided by number of nodes (vertices)
- 5 sinks / leafs / roots in git parlance, nodes with no outgoing edges
- 869 sources / roots / heads in git parlance, nodes with no incoming edge


### The Kotlin language repository

The Kotlin Programming Language  
https://kotlinlang.org/

**Kotlin** is an open-source, statically typed programming language supported and developed by JetBrains and open-source contributors.

In [ ]:
%%time
kotlin_graph = commit_graph('https://github.com/JetBrains/kotlin.git', 'kotlin')

Wall time: 2.24 s


In [ ]:
known_repos.append({
    'url': 'https://github.com/JetBrains/kotlin.git',
    'name': 'kotlin',
    'homepage': 'https://kotlinlang.org/',
    'graph': kotlin_graph,
})
known_repos

[{'name': 'curl',
  'url': 'https://github.com/curl/curl',
  'graph': <networkx.classes.digraph.DiGraph at 0x14d549b1c70>},
 {'name': 'electron',
  'url': 'https://github.com/electron/electron.git',
  'graph': <networkx.classes.digraph.DiGraph at 0x14d549b17f0>},
 {'name': 'ffmpeg',
  'url': 'https://github.com/FFmpeg/FFmpeg.git',
  'graph': <networkx.classes.digraph.DiGraph at 0x14d549b1e50>},
 {'name': 'jgit',
  'url': 'https://github.com/eclipse/jgit.git',
  'graph': <networkx.classes.digraph.DiGraph at 0x14d61fc1280>},
 {'name': 'julia',
  'url': 'https://github.com/JuliaLang/julia.git',
  'homepage': 'https://julialang.org',
  'graph': <networkx.classes.digraph.DiGraph at 0x14d61fc1d30>},
 {'url': 'https://github.com/JetBrains/kotlin.git',
  'name': 'kotlin',
  'homepage': 'https://kotlinlang.org/',
  'graph': <networkx.classes.digraph.DiGraph at 0x14d66efa8e0>}]

GitHub repo @ 01.01.2021: 71,704 commits, 3,007 branches, 20,721 tags

In [ ]:
graph_idx = 5
print('commit graph [{}] of "{}" project has'.format(graph_idx, known_repos[graph_idx]['name']))
print('- {} nodes / vertices / commits'.format(known_repos[graph_idx]['graph'].number_of_nodes()))
print('- {} edges'.format(known_repos[graph_idx]['graph'].number_of_edges()))
print('- {:.5f} edge density, number of edges divided by number of nodes (vertices)'.
      format(known_repos[graph_idx]['graph'].number_of_edges() / known_repos[graph_idx]['graph'].number_of_nodes()))
print('- {} sinks / leafs / roots in git parlance, nodes with no outgoing edges'.
     format(len([n for n in known_repos[graph_idx]['graph'] if known_repos[graph_idx]['graph'].out_degree(n) == 0])))
print('- {} sources / roots / heads in git parlance, nodes with no incoming edge'.
     format(len([n for n in known_repos[graph_idx]['graph'] if known_repos[graph_idx]['graph'].in_degree(n) == 0])))

commit graph [5] of "kotlin" project has
- 117495 nodes / vertices / commits
- 119013 edges
- 1.01292 edge density, number of edges divided by number of nodes (vertices)
- 9 sinks / leafs / roots in git parlance, nodes with no outgoing edges
- 2592 sources / roots / heads in git parlance, nodes with no incoming edge


### Linux repository

Linux kernel source tree (on GitHub)

In [ ]:
%%time
linux_graph = commit_graph('https://github.com/torvalds/linux.git', 'linux')

Wall time: 14.2 s


Wall time: 1min 44s (full clone and extracting the commit graph)

In [ ]:
known_repos.append({
    'url': 'https://github.com/torvalds/linux.git',
    'name': 'linux',
    'graph': linux_graph,
})
known_repos

[{'name': 'curl',
  'url': 'https://github.com/curl/curl',
  'graph': <networkx.classes.digraph.DiGraph at 0x14d549b1c70>},
 {'name': 'electron',
  'url': 'https://github.com/electron/electron.git',
  'graph': <networkx.classes.digraph.DiGraph at 0x14d549b17f0>},
 {'name': 'ffmpeg',
  'url': 'https://github.com/FFmpeg/FFmpeg.git',
  'graph': <networkx.classes.digraph.DiGraph at 0x14d549b1e50>},
 {'name': 'jgit',
  'url': 'https://github.com/eclipse/jgit.git',
  'graph': <networkx.classes.digraph.DiGraph at 0x14d61fc1280>},
 {'name': 'julia',
  'url': 'https://github.com/JuliaLang/julia.git',
  'homepage': 'https://julialang.org',
  'graph': <networkx.classes.digraph.DiGraph at 0x14d61fc1d30>},
 {'url': 'https://github.com/JetBrains/kotlin.git',
  'name': 'kotlin',
  'homepage': 'https://kotlinlang.org/',
  'graph': <networkx.classes.digraph.DiGraph at 0x14d66efa8e0>},
 {'url': 'https://github.com/torvalds/linux.git',
  'name': 'linux',
  'graph': <networkx.classes.digraph.DiGraph at 0x

982,030 commits,  1 branch,  679 tags

In [ ]:
graph_idx = 6
print('commit graph [{}] of "{}" project has'.format(graph_idx, known_repos[graph_idx]['name']))
print('- {} nodes / vertices / commits'.format(known_repos[graph_idx]['graph'].number_of_nodes()))
print('- {} edges'.format(known_repos[graph_idx]['graph'].number_of_edges()))
print('- {:.5f} edge density, number of edges divided by number of nodes (vertices)'.
      format(known_repos[graph_idx]['graph'].number_of_edges() / known_repos[graph_idx]['graph'].number_of_nodes()))
print('- {} sinks / leafs / roots in git parlance, nodes with no outgoing edges'.
     format(len([n for n in known_repos[graph_idx]['graph'] if known_repos[graph_idx]['graph'].out_degree(n) == 0])))
print('- {} sources / roots / heads in git parlance, nodes with no incoming edge'.
     format(len([n for n in known_repos[graph_idx]['graph'] if known_repos[graph_idx]['graph'].in_degree(n) == 0])))

commit graph [6] of "linux" project has
- 982030 nodes / vertices / commits
- 1061095 edges
- 1.08051 edge density, number of edges divided by number of nodes (vertices)
- 4 sinks / leafs / roots in git parlance, nodes with no outgoing edges
- 1 sources / roots / heads in git parlance, nodes with no incoming edge


### Odoo repository

Odoo. Open Source Apps To Grow Your Business.  
<https://www.odoo.com/>

**Odoo** is a suite of web based open source business apps.

In [ ]:
%%time
odoo_graph = commit_graph('https://github.com/odoo/odoo.git', 'odoo')

Wall time: 1.72 s


Wall time 1min 4s (first time `commit_graph()`)

In [ ]:
known_repo = {repo['name']: repo for repo in known_repos}
known_repo

{'curl': {'name': 'curl',
  'url': 'https://github.com/curl/curl',
  'graph': <networkx.classes.digraph.DiGraph at 0x14d549b1c70>},
 'electron': {'name': 'electron',
  'url': 'https://github.com/electron/electron.git',
  'graph': <networkx.classes.digraph.DiGraph at 0x14d549b17f0>},
 'ffmpeg': {'name': 'ffmpeg',
  'url': 'https://github.com/FFmpeg/FFmpeg.git',
  'graph': <networkx.classes.digraph.DiGraph at 0x14d549b1e50>},
 'jgit': {'name': 'jgit',
  'url': 'https://github.com/eclipse/jgit.git',
  'graph': <networkx.classes.digraph.DiGraph at 0x14d61fc1280>},
 'julia': {'name': 'julia',
  'url': 'https://github.com/JuliaLang/julia.git',
  'homepage': 'https://julialang.org',
  'graph': <networkx.classes.digraph.DiGraph at 0x14d61fc1d30>},
 'kotlin': {'url': 'https://github.com/JetBrains/kotlin.git',
  'name': 'kotlin',
  'homepage': 'https://kotlinlang.org/',
  'graph': <networkx.classes.digraph.DiGraph at 0x14d66efa8e0>},
 'linux': {'url': 'https://github.com/torvalds/linux.git',
  '

In [ ]:
known_repo['odoo'] = {
    'name': 'odoo',
    'url': 'https://github.com/odoo/odoo.git',
    'homepage': 'https://www.odoo.com/',
    'graph': odoo_graph,
}
known_repo['odoo']

{'name': 'odoo',
 'url': 'https://github.com/odoo/odoo.git',
 'homepage': 'https://www.odoo.com/',
 'graph': <networkx.classes.digraph.DiGraph at 0x14d549b1d30>}

In [ ]:
repo_name = 'odoo'
print('commit graph of "{}" project has:'.format(repo_name))
print('- {} nodes / vertices / commits'.format(known_repo[repo_name]['graph'].number_of_nodes()))
print('- {} edges'.format(known_repo[repo_name]['graph'].number_of_edges()))
print('- {:.6f} edge density, number of edges divided by number of nodes (vertices)'.
      format(known_repo[repo_name]['graph'].number_of_edges() / known_repo[repo_name]['graph'].number_of_nodes()))
print('- {} sinks / leafs / roots in git parlance, nodes with no outgoing edges'.
     format(len([n for n in known_repo[repo_name]['graph'] if known_repo[repo_name]['graph'].out_degree(n) == 0])))
print('- {} sources / roots / heads in git parlance, nodes with no incoming edge'.
     format(len([n for n in known_repo[repo_name]['graph'] if known_repo[repo_name]['graph'].in_degree(n) == 0])))

commit graph of "odoo" project has:
- 158324 nodes / vertices / commits
- 184403 edges
- 1.164719 edge density, number of edges divided by number of nodes (vertices)
- 7 sinks / leafs / roots in git parlance, nodes with no outgoing edges
- 33 sources / roots / heads in git parlance, nodes with no incoming edge


On GitHub @ 01.01.2021: 139,989 commits,  82 branches,  0 tags

### OpenSSL repository

**OpenSSL** is a robust, commercial-grade, full-featured Open Source Toolkit for the Transport Layer Security (TLS) protocol formerly known as the Secure Sockets Layer (SSL) protocol. The protocol implementation is based on a full-strength general purpose cryptographic library, which can also be used stand-alone.

OpenSSL is descended from the SSLeay library developed by Eric A. Young and Tim J. Hudson.

The official Home Page of the OpenSSL Project is <www.openssl.org>.

In [ ]:
%%time
openssl_graph = commit_graph('https://github.com/openssl/openssl.git', 'openssl')

Wall time: 529 ms


Wall time 18 seconds (first time)

In [ ]:
known_repo['openssl'] = {
    'name': 'openssl',
    'url': 'https://github.com/openssl/openssl.git',
    'homepage': 'https://www.openssl.org',
    'graph': openssl_graph,
}
known_repo['openssl']

{'name': 'openssl',
 'url': 'https://github.com/openssl/openssl.git',
 'homepage': 'https://www.openssl.org',
 'graph': <networkx.classes.digraph.DiGraph at 0x14d569cfbb0>}

In [ ]:
repo_name = 'openssl'
print('commit graph of "{}" project has:'.format(repo_name))
print('- {} nodes / vertices / commits'.format(known_repo[repo_name]['graph'].number_of_nodes()))
print('- {} edges'.format(known_repo[repo_name]['graph'].number_of_edges()))
print('- {:.6f} edge density, number of edges divided by number of nodes (vertices)'.
      format(known_repo[repo_name]['graph'].number_of_edges() / known_repo[repo_name]['graph'].number_of_nodes()))
print('- {} sinks / leafs / roots in git parlance, nodes with no outgoing edges'.
     format(len([n for n in known_repo[repo_name]['graph'] if known_repo[repo_name]['graph'].out_degree(n) == 0])))
print('- {} sources / roots / heads in git parlance, nodes with no incoming edge'.
     format(len([n for n in known_repo[repo_name]['graph'] if known_repo[repo_name]['graph'].in_degree(n) == 0])))

commit graph of "openssl" project has:
- 42818 nodes / vertices / commits
- 42997 edges
- 1.004180 edge density, number of edges divided by number of nodes (vertices)
- 6 sinks / leafs / roots in git parlance, nodes with no outgoing edges
- 19 sources / roots / heads in git parlance, nodes with no incoming edge


OpenSSL repo @ 01.01.2021: 27,812 commits, 20 branches, 320 tags

### The Swift language repository

The Swift Programming Language  
<https://swift.org>

**Swift** is a high-performance system programming language. It has a clean and modern syntax, offers seamless access to existing C and Objective-C code and frameworks, and is memory safe by default.

Although inspired by Objective-C and many other languages, Swift is not itself a C-derived language. As a complete and independent language, Swift packages core features like flow control, data structures, and functions, with high-level constructs like objects, protocols, closures, and generics. Swift embraces modules, eliminating the need for headers and the code duplication they entail.

To learn more about the programming language, visit [swift.org](https://swift.org/documentation/).

In [ ]:
%%time
swift_graph = commit_graph('https://github.com/apple/swift.git', 'swift')

Wall time: 2.65 s


Wall time: 41 s (first time)

In [ ]:
known_repo['swift'] = {
    'name': 'swift',
    'url': 'https://github.com/apple/swift.git',
    'homepage': 'https://swift.org',
    'graph': swift_graph,
    'n_nodes': swift_graph.number_of_nodes(),
    'n_edges': swift_graph.number_of_edges(),
}
known_repo['swift']

{'name': 'swift',
 'url': 'https://github.com/apple/swift.git',
 'homepage': 'https://swift.org',
 'graph': <networkx.classes.digraph.DiGraph at 0x14d61fc1fd0>,
 'n_nodes': 132444,
 'n_edges': 176028}

In [ ]:
repo_name = 'swift'
print('commit graph of "{}" project has:'.format(repo_name))
print('- {} nodes / vertices / commits'.format(known_repo[repo_name]['graph'].number_of_nodes()))
print('- {} edges'.format(known_repo[repo_name]['graph'].number_of_edges()))
print('- {:.6f} edge density, number of edges divided by number of nodes (vertices)'.
      format(known_repo[repo_name]['graph'].number_of_edges() / known_repo[repo_name]['graph'].number_of_nodes()))
print('- {} sinks / leafs / roots in git parlance, nodes with no outgoing edges'.
     format(len([n for n in known_repo[repo_name]['graph'] if known_repo[repo_name]['graph'].out_degree(n) == 0])))
print('- {} sources / roots / heads in git parlance, nodes with no incoming edge'.
     format(len([n for n in known_repo[repo_name]['graph'] if known_repo[repo_name]['graph'].in_degree(n) == 0])))

commit graph of "swift" project has:
- 132444 nodes / vertices / commits
- 176028 edges
- 1.329075 edge density, number of edges divided by number of nodes (vertices)
- 1 sinks / leafs / roots in git parlance, nodes with no outgoing edges
- 209 sources / roots / heads in git parlance, nodes with no incoming edge


Swift repository on GitHub @ 01.01.2021: 114,559 commits

### TensorFlow repository

An Open Source Machine Learning Framework for Everyone  
https://tensorflow.org

**TensorFlow** is an end-to-end open source platform for machine learning. It has a comprehensive, flexible ecosystem of tools, libraries, and community resources that lets researchers push the state-of-the-art in ML and developers easily build and deploy ML-powered applications.

In [ ]:
%%time
tensorflow_graph = commit_graph('https://github.com/tensorflow/tensorflow.git', 'tensorflow')

Wall time: 1.15 s


Wall time: 34.3 s (first time)

In [ ]:
known_repo['tensorflow'] = {
    'name': 'tensorflow',
    'url': 'https://github.com/tensorflow/tensorflow.git',
    'homepage': 'https://tensorflow.org',
    'graph': tensorflow_graph,
    'n_nodes': tensorflow_graph.number_of_nodes(),
    'n_edges': tensorflow_graph.number_of_edges(),
}
known_repo['tensorflow']

{'name': 'tensorflow',
 'url': 'https://github.com/tensorflow/tensorflow.git',
 'homepage': 'https://tensorflow.org',
 'graph': <networkx.classes.digraph.DiGraph at 0x14d549b1490>,
 'n_nodes': 105236,
 'n_edges': 115687}

In [ ]:
repo_name = 'tensorflow'
print('commit graph of "{}" project has:'.format(repo_name))
print('- {} nodes / vertices / commits'.format(known_repo[repo_name]['n_nodes']))
print('- {} edges'.format(known_repo[repo_name]['n_edges']))
print('- {:.6f} edge density, number of edges divided by number of nodes (vertices)'.
      format(known_repo[repo_name]['n_edges'] / known_repo[repo_name]['n_nodes']))
print('- {} sinks / leafs / roots in git parlance, nodes with no outgoing edges'.
     format(len([n for n in known_repo[repo_name]['graph'] if known_repo[repo_name]['graph'].out_degree(n) == 0])))
print('- {} sources / roots / heads in git parlance, nodes with no incoming edge'.
     format(len([n for n in known_repo[repo_name]['graph'] if known_repo[repo_name]['graph'].in_degree(n) == 0])))

commit graph of "tensorflow" project has:
- 105236 nodes / vertices / commits
- 115687 edges
- 1.099310 edge density, number of edges divided by number of nodes (vertices)
- 3 sinks / leafs / roots in git parlance, nodes with no outgoing edges
- 32 sources / roots / heads in git parlance, nodes with no incoming edge


TensorFlow repository on GitHub @ 01.01.2021: 102,131 commits,  36 branches

### The TypeScript language repository

TypeScript is a superset of JavaScript that compiles to clean JavaScript output.  
www.typescriptlang.org

**TypeScript** is a language for application-scale JavaScript. TypeScript adds optional types to JavaScript that support tools for large-scale JavaScript applications for any browser, for any host, on any OS. TypeScript compiles to readable, standards-based JavaScript.

In [ ]:
%%time
typescript_graph = commit_graph('https://github.com/microsoft/TypeScript.git', 'typescript')

Wall time: 397 ms


Wall time: 23.4 s (first time)

In [ ]:
known_repo['typescript'] = {
    'name': 'typescript',
    'url': 'https://github.com/microsoft/TypeScript.git',
    'homepage': 'https://www.typescriptlang.org/',
    'graph': typescript_graph,
    'n_nodes': typescript_graph.number_of_nodes(),
    'n_edges': typescript_graph.number_of_edges(),
}
known_repo['typescript']

{'name': 'typescript',
 'url': 'https://github.com/microsoft/TypeScript.git',
 'homepage': 'https://www.typescriptlang.org/',
 'graph': <networkx.classes.digraph.DiGraph at 0x14d66efad90>,
 'n_nodes': 35486,
 'n_edges': 43895}

In [ ]:
repo_name = 'typescript'
print('commit graph of "{}" project has:'.format(repo_name))
print('- {} nodes / vertices / commits'.format(known_repo[repo_name]['n_nodes']))
print('- {} edges'.format(known_repo[repo_name]['n_edges']))
print('- {:.6f} edge density, number of edges divided by number of nodes (vertices)'.
      format(known_repo[repo_name]['n_edges'] / known_repo[repo_name]['n_nodes']))
print('- {} sinks / leafs / roots in git parlance, nodes with no outgoing edges'.
     format(len([n for n in known_repo[repo_name]['graph'] if known_repo[repo_name]['graph'].out_degree(n) == 0])))
print('- {} sources / roots / heads in git parlance, nodes with no incoming edge'.
     format(len([n for n in known_repo[repo_name]['graph'] if known_repo[repo_name]['graph'].in_degree(n) == 0])))

commit graph of "typescript" project has:
- 35486 nodes / vertices / commits
- 43895 edges
- 1.236967 edge density, number of edges divided by number of nodes (vertices)
- 1 sinks / leafs / roots in git parlance, nodes with no outgoing edges
- 419 sources / roots / heads in git parlance, nodes with no incoming edge


TypeScript repository on GitHub @ 01.01.2021: 31,596 commits, 455 branches

### Git project repository

<https://github.com/git/git> is Git Source Code Mirror - This is a publish-only repository and all pull requests are ignored.  
https://git-scm.com/

**Git** is a fast, scalable, distributed revision control system with an unusually rich command set that provides both high-level operations and full access to internals.

In [ ]:
%%time
git_graph = commit_graph('https://github.com/git/git.git', 'git.git')

Wall time: 824 ms


In [ ]:
known_repo['git.git'] = {
    'name': 'git.git',
    'graph_name': _repo_graph_name('repos/git.git'),
    'url': 'https://github.com/git/git.git',
    'homepage': 'https://git-scm.com/',
    'graph': git_graph,
    'n_nodes': git_graph.number_of_nodes(),
    'n_edges': git_graph.number_of_edges(),
}
known_repo['git.git']

{'name': 'git.git',
 'graph_name': 'git-commit_graph',
 'url': 'https://github.com/git/git.git',
 'homepage': 'https://git-scm.com/',
 'graph': <networkx.classes.digraph.DiGraph at 0x14d66efa2e0>,
 'n_nodes': 63829,
 'n_edges': 79664}

In [ ]:
repo_name = 'git.git'
print('commit graph of "{}" project has:'.format(repo_name))
print('- {} nodes / vertices / commits'.format(known_repo[repo_name]['n_nodes']))
print('- {} edges'.format(known_repo[repo_name]['n_edges']))
print('- {:.6f} edge density, number of edges divided by number of nodes (vertices)'.
      format(known_repo[repo_name]['n_edges'] / known_repo[repo_name]['n_nodes']))
print('- {} sinks / leafs / roots in git parlance, nodes with no outgoing edges'.
     format(len([n for n in known_repo[repo_name]['graph'] if known_repo[repo_name]['graph'].out_degree(n) == 0])))
print('- {} sources / roots / heads in git parlance, nodes with no incoming edge'.
     format(len([n for n in known_repo[repo_name]['graph'] if known_repo[repo_name]['graph'].in_degree(n) == 0])))

commit graph of "git.git" project has:
- 63829 nodes / vertices / commits
- 79664 edges
- 1.248085 edge density, number of edges divided by number of nodes (vertices)
- 9 sinks / leafs / roots in git parlance, nodes with no outgoing edges
- 3 sources / roots / heads in git parlance, nodes with no incoming edge


Git.git repository on GitHub @ 01.01.2021: 61,475 commits,  5 branches

### Gerrit repository

**Gerrit Code Review** is a free, web-based team code collaboration tool. Software developers in a team can review each other's modifications on their source code using a Web browser and approve or reject those changes. It integrates closely with Git, a distributed version control system.

https://www.gerritcodereview.com/

In [ ]:
%%time
gerrit_graph = commit_graph('https://gerrit.googlesource.com/gerrit', 'gerrit')

Wall time: 580 ms


Wall time: 1min 2s (first time)

In [ ]:
%%time
known_repo['gerrit'] = {
    'name': 'gerrit',
    'graph_name': _repo_graph_name('repos/gerrit'),
    'url': 'https://gerrit.googlesource.com/gerrit',
    'homepage': 'https://www.gerritcodereview.com/',
    'graph': gerrit_graph,
    'n_nodes': gerrit_graph.number_of_nodes(),
    'n_edges': gerrit_graph.number_of_edges(),
}
known_repo['gerrit']

Wall time: 68 ms


{'name': 'gerrit',
 'graph_name': 'gerrit-commit_graph',
 'url': 'https://gerrit.googlesource.com/gerrit',
 'homepage': 'https://www.gerritcodereview.com/',
 'graph': <networkx.classes.digraph.DiGraph at 0x14d9516e4f0>,
 'n_nodes': 47687,
 'n_edges': 63944}

In [ ]:
repo_name = 'gerrit'
print('commit graph of "{}" project has:'.format(repo_name))
print('- {} nodes / vertices / commits'.format(known_repo[repo_name]['n_nodes']))
print('- {} edges'.format(known_repo[repo_name]['n_edges']))
print('- {:.6f} edge density, number of edges divided by number of nodes (vertices)'.
      format(known_repo[repo_name]['n_edges'] / known_repo[repo_name]['n_nodes']))
print('- {} sinks / leafs / roots in git parlance, nodes with no outgoing edges'.
     format(len([n for n in known_repo[repo_name]['graph'] if known_repo[repo_name]['graph'].out_degree(n) == 0])))
print('- {} sources / roots / heads in git parlance, nodes with no incoming edge'.
     format(len([n for n in known_repo[repo_name]['graph'] if known_repo[repo_name]['graph'].in_degree(n) == 0])))

commit graph of "gerrit" project has:
- 47687 nodes / vertices / commits
- 63944 edges
- 1.340911 edge density, number of edges divided by number of nodes (vertices)
- 4 sinks / leafs / roots in git parlance, nodes with no outgoing edges
- 6 sources / roots / heads in git parlance, nodes with no incoming edge


### Android-base repository

In [ ]:
%%time
android_graph = commit_graph('https://android.googlesource.com/platform/frameworks/base/', 'android-base', rescan=True)

Wall time: 17.5 s


Wall time: 2min 9s (first time), 20.7 s (rescan)

In [ ]:
#%%time
#!git clone --filter=tree:0 --no-checkout https://android.googlesource.com/platform/frameworks/base/ repos/android-base

In [ ]:
known_repo['android-base'] = {
    'name': 'android-base',
    'url': 'https://android.googlesource.com/platform/frameworks/base/',
    'graph': android_graph,
    'n_nodes': android_graph.number_of_nodes(),
    'n_edges': android_graph.number_of_edges(),
}
known_repo['android-base']

{'name': 'android-base',
 'url': 'https://android.googlesource.com/platform/frameworks/base/',
 'graph': <networkx.classes.digraph.DiGraph at 0x14d9516e9a0>,
 'n_nodes': 559423,
 'n_edges': 984070}

In [ ]:
repo_name = 'android-base'
print('commit graph of "{}" project has:'.format(repo_name))
print('- {} nodes / vertices / commits'.format(known_repo[repo_name]['n_nodes']))
print('- {} edges'.format(known_repo[repo_name]['n_edges']))
print('- {:.6f} edge density, number of edges divided by number of nodes (vertices)'.
      format(known_repo[repo_name]['n_edges'] / known_repo[repo_name]['n_nodes']))
print('- {} sinks / leafs / roots in git parlance, nodes with no outgoing edges'.
     format(len([n for n in known_repo[repo_name]['graph'] if known_repo[repo_name]['graph'].out_degree(n) == 0])))
print('- {} sources / roots / heads in git parlance, nodes with no incoming edge'.
     format(len([n for n in known_repo[repo_name]['graph'] if known_repo[repo_name]['graph'].in_degree(n) == 0])))

commit graph of "android-base" project has:
- 559423 nodes / vertices / commits
- 984070 edges
- 1.759080 edge density, number of edges divided by number of nodes (vertices)
- 3 sinks / leafs / roots in git parlance, nodes with no outgoing edges
- 1 sources / roots / heads in git parlance, nodes with no incoming edge


### Chromium repository

**Chromium** is an open-source browser project that aims to build a safer, faster, and more stable way for all users to experience the web.

The project's web site is https://www.chromium.org.

In [ ]:
%%time
chromium_graph = commit_graph('https://chromium.googlesource.com/chromium/src', 'chromium')

Wall time: 39.3 s


Wall time: 39min 52s (first time, possibly failed), rescan: 39.3 s

In [ ]:
known_repo['chromium'] = {
    'name': 'chromium',
    'url': 'https://chromium.googlesource.com/chromium/src',
    'homepage': 'https://www.chromium.org',
    'graph': chromium_graph,
    'n_nodes': chromium_graph.number_of_nodes(),
    'n_edges': chromium_graph.number_of_edges(),
}
known_repo['chromium']

{'name': 'chromium',
 'url': 'https://chromium.googlesource.com/chromium/src',
 'homepage': 'https://www.chromium.org',
 'graph': <networkx.classes.digraph.DiGraph at 0x14d549b16d0>,
 'n_nodes': 961204,
 'n_edges': 961263}

In [ ]:
repo_name = 'chromium'
print('commit graph of "{}" project has:'.format(repo_name))
print('- {} nodes / vertices / commits'.format(known_repo[repo_name]['n_nodes']))
print('- {} edges'.format(known_repo[repo_name]['n_edges']))
print('- {:.6f} edge density, number of edges divided by number of nodes (vertices)'.
      format(known_repo[repo_name]['n_edges'] / known_repo[repo_name]['n_nodes']))
print('- {} sinks / leafs / roots in git parlance, nodes with no outgoing edges'.
     format(len([n for n in known_repo[repo_name]['graph'] if known_repo[repo_name]['graph'].out_degree(n) == 0])))
print('- {} sources / roots / heads in git parlance, nodes with no incoming edge'.
     format(len([n for n in known_repo[repo_name]['graph'] if known_repo[repo_name]['graph'].in_degree(n) == 0])))

commit graph of "chromium" project has:
- 961204 nodes / vertices / commits
- 961263 edges
- 1.000061 edge density, number of edges divided by number of nodes (vertices)
- 3 sinks / leafs / roots in git parlance, nodes with no outgoing edges
- 1 sources / roots / heads in git parlance, nodes with no incoming edge


## Summary

Imports for data analysis

In [ ]:
import pandas as pd

In [ ]:
for name, repo in known_repo.items():
    print('... processing {}'.format(name))
    if 'n_nodes' not in repo:
        repo['n_nodes'] = repo['graph'].number_of_nodes()
    if 'n_edges' not in repo:
        repo['n_edges'] = repo['graph'].number_of_edges()

... processing curl
... processing electron
... processing ffmpeg
... processing jgit
... processing julia
... processing kotlin
... processing linux
... processing odoo
... processing openssl
... processing swift
... processing tensorflow
... processing typescript
... processing git.git
... processing gerrit
... processing android-base
... processing chromium


In [ ]:
known_repo_df = pd.DataFrame.from_dict(known_repo, orient='index',
                                       columns=['url', 'n_nodes', 'n_edges']).astype({'url': 'str'})
known_repo_df['n_edges / n_nodes'] = known_repo_df['n_edges'] / known_repo_df['n_nodes']
known_repo_df.sort_values('n_nodes')

url  n_nodes  \
jgit                        https://github.com/eclipse/jgit.git     8173   
curl                               https://github.com/curl/curl    26609   
electron               https://github.com/electron/electron.git    29956   
typescript          https://github.com/microsoft/TypeScript.git    35486   
openssl                  https://github.com/openssl/openssl.git    42818   
gerrit                   https://gerrit.googlesource.com/gerrit    47687   
julia                    https://github.com/JuliaLang/julia.git    56048   
git.git                          https://github.com/git/git.git    63829   
tensorflow         https://github.com/tensorflow/tensorflow.git   105236   
kotlin                  https://github.com/JetBrains/kotlin.git   117495   
ffmpeg                     https://github.com/FFmpeg/FFmpeg.git   120583   
swift                        https://github.com/apple/swift.git   132444   
odoo                           https://github.com/odoo/odoo.git   158324   
android-base  https://android.googlesource.com/platform/fram...   559423   
chromium         https://chromium.googlesource.com/chromium/src   961204   
linux                     https://github.com/torvalds/linux.git   982030   

              n_edges  n_edges / n_nodes  
jgit             9853           1.205555  
curl            26631           1.000827  
electron        34167           1.140573  
typescript      43895           1.236967  
openssl         42997           1.004180  
gerrit          63944           1.340911  
julia           67547           1.205163  
git.git         79664           1.248085  
tensorflow     115687           1.099310  
kotlin         119013           1.012920  
ffmpeg         130606           1.083121  
swift          176028           1.329075  
odoo           184403           1.164719  
android-base   984070           1.759080  
chromium       961263           1.000061  
linux         1061095           1.080512

In [ ]:
known_repo_df.dtypes

url                   object
n_nodes                int64
n_edges                int64
n_edges / n_nodes    float64
dtype: object